### PR TITLE
Add cross-dataset class-row weight transfer for RT-DETR

### PR DESCRIPTION
**Pull request by artest08**
## 🛠️ Summary

Enables RT-DETR fine-tuning from a checkpoint pretrained on a different
dataset (e.g. Objects365 → COCO) without discarding the classification
head. Class-row tensors in `enc_score_head` and `dec_score_head` are
remapped from source to target by class-name matching; rows for
target classes that exist in the source are copied, rows for target-only
classes keep their random init.

This is a common technique in the literature for boosting downstream
mAP when the pretraining dataset has a class-name overlap with the
target. The implementation is generic — any name-aligned dataset pair
benefits, not just O365 → COCO.

## 🎯 Motivation

Today `model.load()` falls back to `intersect_dicts`, which silently
drops every classification-head tensor when `nc` differs. Users cannot benefit from the learnt classification heads even when class semantics are preserved in the source. This PR closes that gap.

## 📦 Changes

- **New**: `ultralytics/utils/class_map.py` — pure-tensor helpers for
  name normalization, class-row index mapping, and state-dict remap.
- `ultralytics/nn/tasks.py` — `RTDETRDetectionModel.load` now:
  - drops `denoising_class_embed` on shape mismatch (it's an Embedding
    table, can't be partially transferred), and
  - remaps class-row tensors by name when both src and dst names are
    available and non-numeric.
- `ultralytics/models/rtdetr/train.py` — passes `weights.names` and
  `self.data["names"]` through to `model.load()`.

## 🔍 How it works

1. Source/target class-name lists are normalized (lowercase, strip
   punctuation, collapse whitespace).
2. A small alias dict handles known cross-dataset spelling differences
   (e.g. `bird ↔ wild bird`) used **only** when direct match fails.
3. For every state-dict key matching `score_head` or `class_embed`
   with `shape[1:]` agreement, rows are copied src→dst per the name
   map. Unmapped target rows keep their pre-load init.
4. Numeric-placeholder name lists (e.g. `["0","1",...]`) skip remap
   entirely so legacy checkpoints behave as before.

## ✅ Testing

Smoke-tested end-to-end on `coco8` with a 24-epoch O365-pretrained
YOLO26L-RTDETR checkpoint:

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
🎯 RT-DETR pretrained classification weights are now remapped by class name, improving transfer when datasets use different class orders or counts.

### 📊 Key Changes
- Added RT-DETR-specific class-head remapping for `score_head` and `class_embed` tensors.
- Transfers matching class rows by name, regardless of source and target class order.
- Preserves target initialization for classes missing from the pretrained model.
- Supports partial transfer of the denoising class embeddings.
- Integrated dataset class names into RT-DETR model loading.
- Added tests covering reordered classes, unmatched classes, and denoising embeddings.
- Added type annotations and future-annotations support in `tasks.py`.

### 🎯 Purpose & Impact
- ✅ Makes pretrained RT-DETR weights more useful when fine-tuning on datasets with different class definitions.
- 🔄 Prevents class predictions from being assigned to the wrong labels when class ordering changes.
- ➕ Allows new target classes to retain their initialized weights instead of receiving incorrect pretrained values.
- 🧪 Improves loading reliability while maintaining existing behavior when class names already match.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @artest08, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with review readiness, and an engineer will assist you shortly. Please review the checklist below for a smooth integration of your RT-DETR update 🔍✨

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/utils/class_map.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+    row_map, missing = build_class_row_map(src_name_list, dst_name_list, aliases=aliases)
+    remapped = dict(src_state)
+    remapped_keys: list[tuple[str, tuple[int, ...], tuple[int, ...]]] = []
+
+    for key, src_tensor in src_state.items():
+        if key not in dst_state:
+            continue
+        dst_tensor = dst_state[key]
+        if src_tensor.shape == dst_tensor.shape:
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/class_map.py]: ⚠️ **HIGH**: This skips remapping whenever the source and destination tensors have the same shape. That means datasets with the same class count but a different class order will still load the pretrained classification rows unchanged, silently assigning weights to the wrong classes. The remap should run whenever the class-name mapping is non-identity, not only on shape mismatches.

File: ultralytics/utils/class_map.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+) -> tuple[dict[int, int], list[tuple[int, str]]]:
+    """Build destination-to-source class row mapping from class-name lists."""
+    aliases = aliases or DEFAULT_CLASS_ALIASES
+    src_norm_to_idx = {normalize_class_name(name): i for i, name in enumerate(src_names)}
+
+    row_map: dict[int, int] = {}
+    missing: list[tuple[int, str]] = []
+    for dst_idx, dst_name in enumerate(dst_names):
+        key = normalize_class_name(dst_name)
+        alias_values = aliases.get(key, [])
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/class_map.py]: 💡 **MEDIUM**: The alias lookup is one-way: it only checks aliases for the destination class name before searching the source names. With the current table (`"bird" -> "wild bird"`, etc.), transfer works only when the target uses the canonical key and the source uses the alias, but fails in the reverse direction. Since source/target can be either dataset, this makes remapping asymmetric and will miss valid matches for the opposite transfer direction.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The overall approach looks solid and the RT-DETR loading flow is integrated cleanly, but there are two important remapping logic gaps to address before merging: same-size class heads are not reordered, and alias matching is currently asymmetric across source/target datasets.

💬 Posted 2 inline comments

**Comment by Laughing-q** [comment]
@artest08 do you think we can also have the similar feature for YOLO models as well?

**Comment by artest08** [comment]
@Laughing-q 

I opened a branch, `yolo-class-weight-transfer`, and applied the same class-weight projection logic for YOLO detection models.

As a POC, we can try training `yolo26s.yaml` initialized from `yolo26s-objv1-150.pt`. With `DEFAULT_CLASS_ALIASES` set to:

```python
"bird": "wild bird",
"sports ball": "baseball",
```

all 80 COCO classes are initialized from pretrained class weights. Without these aliases, 78/80 classes are matched directly by name.

For `yolo26s`, the load looks clean:

```text
Transferred 708/708 items from pretrained weights
```

So this seems like a good candidate for a quick training comparison. You know better than me the best hyperparameters for YOLO models therefore you can start a yolo26s training to observe whether there is any improvement with this implementation?

**Comment by artest08** [comment]
What I believe is that this structure does not ensure higher mAP values but what it is ensuring is faster mAP convergence, I think.

**Comment by Laughing-q** [comment]
@Y-T-G

**Comment by glenn-jocher** [comment]
Hi @artest08 — I merged the latest `main` into your branch to clear the `docs/mkdocs_github_authors.yaml` conflict, so it's current again and CI can run clean. I also re-reviewed the effective diff and the class-row remap logic holds up: I unit-tested `remap_class_row_state_dict` on a mismatched src/dst pair and confirmed matched classes copy the correct rows (both weight and bias) while unmatched target classes keep their init, and `is_default_numeric_names` correctly gates legacy numeric-name checkpoints so they behave as before.

Two concrete items before this can merge:

1. **Docs frontmatter placeholders** — `docs/en/reference/utils/class_map.md` still carries `description: TODO ADD DESCRIPTION` and `keywords: TODO ADD KEYWORDS`, which would publish verbatim to docs.ultralytics.com. Please fill those in (or delete the file and re-run `python docs/build_reference.py` for sensible defaults).

2. **A small unit test** — Codecov flags `class_map.py` at ~17% patch coverage. A quick test in an existing `tests/*.py` exercising `remap_class_row_state_dict` (one matched row copied, one unmatched row kept) plus `is_default_numeric_names` would satisfy the checklist and lock the behavior in against regressions.

Minor note: the new `RTDETRDetectionModel.load` override drops the base `load`'s `model.0.conv.weight` multi-channel transfer branch. It's a no-op for the standard HGNetv2 RT-DETR backbone, but for the YOLO-backbone RT-DETR variants it would skip the partial first-conv copy during multi-channel (`ch != 3`) training — worth preserving if you expect that path.

Nice feature overall — reusing name-aligned classification rows for faster convergence is a genuine win.

**Comment by artest08** [comment]
Hi @glenn-jocher, thanks for the valuable feedback. Refinements pushed:

- Filled in `class_map.md` description + keywords.
- Added `test_utils_class_map` covering `is_default_numeric_names` and `remap_class_row_state_dict` (matched, alias-driven, unmatched).
- Preserved the `model.0.conv.weight` multi-channel branch via a shared `BaseModel._transfer_state_dict` helper.

CC: @Laughing-q

**Comment by artest08** [comment]
Hi @glenn-jocher, the PR is now aligned with the recently-merged YOLO cls-head name-remap. After the main merge:

- RTDETRDetectionModel._remap_cls_by_names override for RT-DETR's score_head/class_embed layout + denoising_class_embed discard on nc mismatch
- RTDETRTrainer.get_model wraps construction in set_model_names_for_load (same pattern as YOLO detect/pose/seg/obb trainers)
- Removed the earlier standalone ultralytics/utils/class_map.py and its docs entry (superseded by the new extension point)

Cross-dataset aliases are moved out of this PR; I'll open a follow-up with a userland example to the documentation. 

CC: @Laughing-q

**Review by Laughing-q — APPROVED** [review]

**Comment by Laughing-q** [comment]
@artest08 merging! nice work!

**Comment by UltralyticsAssistant** [comment]
🎉 Thank you for the excellent contribution, @artest08, with support from @glenn-jocher and @Laughing-q! This PR’s RT-DETR class-head remapping makes pretrained weights far more reliable and useful across datasets with reordered, expanded, or partially overlapping class definitions—helping prevent incorrect label assignments while preserving sensible initialization for new classes.

As Helen Keller said, “Alone we can do so little; together we can do so much.” Your collaboration turns a subtle transfer-learning challenge into a robust improvement for the Ultralytics community. 🙌
